### PR TITLE
Add support to replace an item in list

### PR DESCRIPTION
**Pull request by iamorchid**
Hi adamw,

I'm quite interested in your quicklens. Here I create a PR to implement replacing a specific item in the list. And for the usage of this functionality, you could refer to the unittest ModifyAtTest.scala

Thanks,
Will

### CODE REVIEW

**Comment by benhutchison** [comment]
Im guessing this PR arose from issue #9 ? 

Im just a user of quicklens, but Im concerned that this adds a feature that can be reasonably achieved without specific support in the library. 

`SeqLike` already provides [updated](http://www.scala-lang.org/api/2.11.5/index.html#scala.collection.SeqLike@updated(index:Int,elem:A):Seq[A]) and a lens-like `def update(index: Int, f: A => A): Seq[A]` can be pimped onto Seq by anyone who needs it. 

What situations require support within quicklens particularly? If the `f` in `update` needs to copy nested fields, it's implementation can use quicklens Eg assuming `update` we can do:
```
case class Person(name: String, age: Int)
Vector(Person("Ben", 41)).update(0, _.modify(_.age).using(_ + 1))
```

Is the lack of something like `def update(index: Int, f: A => A): Seq[A]` on Seq by default the real problem? IMHO I dont think that's quicklens' problem to solve..

**Comment by iamorchid** [comment]
Yea, for the simple case you offer here, the feature in my PR seems not so useful. What if the following case ?
``` scala
case class A(name: String)
case class B(a: A)
case class C(bs: List[B])
case class D(cs: List[C])
case class E(d: D)
```
Given an instance e of class E, how do you update ```e.d.cs(1).bs(2).a.name``` in your case ? It would looks quite complicated and not idiomatic to quicklens, right ? However, with the new feature, we could use ```modify(e)(_.d.cs.at(1).at(2).a.name).using(duplicate)```.

File: quicklens/src/main/scala/com/softwaremill/quicklens/package.scala [hunk 1]
```diff
@@ -47,20 +48,43 @@ package object quicklens {
     }
   }
 
-  implicit class QuicklensEach[F[_], T](t: F[T])(implicit f: QuicklensFunctor[F, T, T]) {
+  implicit class QuicklensEach[F[_], T](t: F[T])(implicit f: QuicklensEachFunctor[F, T]) {
```
Comment (adamw) [inline quicklens/src/main/scala/com/softwaremill/quicklens/package.scala]: is this change necessary? Maybe not used now, but the functors can be kept general I suppose?

File: quicklens/src/main/scala/com/softwaremill/quicklens/package.scala [hunk 2]
```diff
@@ -47,20 +48,43 @@ package object quicklens {
...
+  trait QuicklensAtFunctor[F[_], T] {
+    def at(fa: F[T], idx: Int)(f: T => T): F[T]
+  }
+
+  implicit def seqQuicklensFunctor[F[_], T](implicit cbf: CanBuildFrom[F[T], T, F[T]], ev: F[T] => SeqLike[T, F[T]]) =
+    new QuicklensAtFunctor[F, T] {
+      override def at(fa: F[T], idx: Int)(f: T => T) = {
+        val builder = cbf(fa)
+        for (i <- 0 to (fa.size - 1)) {
+          if (i == idx) builder += f(fa(idx))
```
Comment (adamw) [inline quicklens/src/main/scala/com/softwaremill/quicklens/package.scala]: This is going to be slow e.g. for `List`. Maybe we could use an ugly `.foreach` + a manually maintained counter? (incremented in the foreach body)

**Comment by adamw** [comment]
Looks good overall - thanks! As it's not the first request for such a feature I think adding it is a good idea. Just two minor remarks :)

File: quicklens/src/main/scala/com/softwaremill/quicklens/package.scala (hunk 1 above)
Comment (iamorchid) [inline quicklens/src/main/scala/com/softwaremill/quicklens/package.scala]: You're right. We could keep the generic one for future extension.

File: quicklens/src/main/scala/com/softwaremill/quicklens/package.scala (hunk 2 above)
Comment (iamorchid) [inline quicklens/src/main/scala/com/softwaremill/quicklens/package.scala]: Not quite understand why using for-loop here is slower than native foreach.

Comment (adamw) [inline quicklens/src/main/scala/com/softwaremill/quicklens/package.scala]: Because at each iteration you query the collection at the current index: `fa(idx)`. In case of list, that can take linear time

Comment (iamorchid) [inline quicklens/src/main/scala/com/softwaremill/quicklens/package.scala]: I see. Thanks.

**Comment by adamw** [comment]
1.4.1 (with some changes) released :)
